q2_types/feature_data/_format.py

```diff
@@ -131,23 +136,82 @@ def sniff(self):
+                # If we read 2 and only get one \n we have only one \n in the
+                # file
+                first = fh.read(2)
+                # Empty files should pass
+                if first == '\n' or first == '':
+                    return
+                if first[0] != '>':
+                    raise ValidationError(
+                        "First line of file is not a valid FASTA ID. FASTA "
+                        "IDs must start with '>'")
+                fh.seek(0)
+                line = fh.readline()
+                while line != '' and line_number < level:
+                    if line.startswith('>'):
+                        if last_line_was_ID:
+                            raise ValidationError(
+                                'Multiple consecutive IDs starting on line '
+                                f'{line_number-1!r}')
+                        last_line_was_ID = True
+                    elif re.fullmatch(FASTADNAValidator, line):
+                        last_line_was_ID = False
+                    else:
+                        raise ValidationError(
+                            f'Invalid sequence on line {line_number}')
+                    line_number += 1
+                    line = fh.readline()
+            except UnicodeDecodeError as e:
+                # We tell() on the buffer because we want the actual buffer the
+                # error occured in, tell() on the file handle will report being
+                # at the end of the last buffer when the error occured well
+                # into the next buffer
+                buffer = fh.buffer
+                pos = buffer.tell()
+                # We want to start our read from the beginning of the buffer we
+                # encountered the error on because we have yet to count any
+                # lines in that buffer
+                pos = (math.ceil(pos / io.DEFAULT_BUFFER_SIZE) - 1) * \
+                    io.DEFAULT_BUFFER_SIZE
+                buffer.seek(pos)
+                # e.start reports the position of the bad byte in the current
+                # buffer not the file, this is the cause of all the pain
+                # determining which buffer in the file the error occured in so
+                # we can find the bad byte's position in the file not just the
+                # buffer. If the bad byte is the final byte in a buffer it will
+                # report it as being the first byte in the next buffer. This is
+                # likely due to the fact that e.start reports the buffer
+                # position of the bad byte, and e.end reports the buffer
+                # position of the next good byte. If the last byte in a buffer
+                # is bad then e.start should be 8192 and e.end should be byte 0
+                # of the next buffer, but end being less than start doesn't
+                # really make sense, so this is handled by bumping e.start to 0
+                # and e.end to 1
```

Comment:
I may be mistaken, but the behavior described doesn't _seem_ quite right.

I think the interval described in this comment may be off by one (the code is fine), it sounds as though `e.start` is the bad byte, but isn't it actually the byte immediately preceding the bad byte? i.e. the bad bytes are something like this:
`(e.start, e.end)` instead of `[e.start, e.end)`?

Reply:
Start is ["The first index of invalid data in object."](https://docs.python.org/3.6/library/exceptions.html#UnicodeError) and end is "The index after the last invalid data in object. " Object is the current buffer, so it should be 8191 not 8192 because 0 based indexing, but that index is the first bad byte

Reply:
Fair enough! And you've confirmed the e.end behavior does indeed wrap around as described (setting start to 0)?

Reply:
Yes, if the last byte of a buffer is invalid it reports e.start=0 and e.end=1 just the same as if the first byte in a buffer is invalid